### PR TITLE
Add `Compression` middleware

### PR DESCRIPTION
**Pull request by davidpdrsn**
Suggested by @LucioFranco in https://github.com/tower-rs/tower-http/pull/33:

> I think this is fine for now but I want to consider if we could maybe impl this over impl Body<Buf = Bytes> instead of hyper::Body. This way we could eventually add this into tonic quite easily. That is because in tonic we impl body like so https://docs.rs/tonic/0.4.0/tonic/body/struct.BoxBody.html#impl-Body

I totally agree but I couldn't quite figure out how to make it work. `Body::wrap_stream` is quite convenient. I could give it another shot.

## TODO

- [x] Add feature toggles for each compression algorithm a la https://github.com/tower-rs/tower-http/pull/41/commits/090053ae3056996e29228f8a93739227c16d01ab#r566686758.

### CODE REVIEW

**Comment by davidpdrsn** [comment]
I've put this back into draft mode. https://github.com/tower-rs/tower-http/pull/41 shows how to make decompression work without depending on `hyper::Body`. So I'm gonna copy that approach for this PR.

**Comment by davidpdrsn** [comment]
I've updated this to support any `B: http_body::Body` rather than being hardcoded to `hyper::Body`. Its highly inspired by https://github.com/tower-rs/tower-http/pull/41 so thanks to @tesaguri!

Before merging this we should decide something regarding the questions raised [here](https://github.com/tower-rs/tower-http/pull/41#pullrequestreview-577107333) though.

File: tower-http/src/compression.rs
Comment (LucioFranco) [inline tower-http/src/compression.rs:272]: I wonder if there is a way to optimize this if it impls bytes :(

File: tower-http/src/compression.rs
Comment (LucioFranco) [inline tower-http/src/compression.rs:214]: lol

**Review by LucioFranco — APPROVED** [review]
This looks great!

File: tower-http/src/compression.rs
Comment (tesaguri) [reply tower-http/src/compression.rs:272]: I think `<Bytes as Buf>::copy_to_bytes` is already optimized enough. It will become a mere `mem::replace` call if the argument equals `data.remaining()`, without even incrementing the reference count.

<https://github.com/tokio-rs/bytes/blob/master/src/bytes.rs#L551-L559>

**Pull request merged**
